File: docs/palindromes-in-peg/scaffold_rom.py
"""Finite lookup tables as shared Boolean decision diagrams.

The table is construction data. A read instantiates its fixed Boolean DAG
on the current finite address bits; there is no runtime array access or
semantic callback in the emitted scaffold/PEG.
"""
from scaffold_circuit import Value, TRUE, FALSE, choose


class ROM:
  def __init__(self, rows, domains):
    self.rows = tuple(dict(row) for row in rows)
    self.domains = {key: tuple(domain) for key, domain in domains.items()}
    if not self.rows or any(set(row) != set(domains) for row in self.rows):
      raise ValueError("every finite table row must cover the declared columns")
    if any(not domain or len(set(domain)) != len(domain) for domain in self.domains.values()):
      raise ValueError("finite column values must be distinct")
    self.width = max(0, (len(self.rows) - 1).bit_length())
    self.nodes, unique, memo = [], {}, {}

    def build(values, bit):
      if not any(values): return 0
      if all(values): return 1
      key = bit, values
      if key in memo: return memo[key]
      half = len(values) // 2
      no, yes = build(values[:half], bit - 1), build(values[half:], bit - 1)
      if yes == no:
        result = yes
      else:
        node = bit, yes, no
        if node not in unique:
          unique[node] = len(self.nodes) + 2
          self.nodes.append(node)
        result = unique[node]
      memo[key] = result
      return result

    self.outputs = {}
    padded = self.rows + (self.rows[0],) * ((1 << self.width) - len(self.rows))
    for key, domain in self.domains.items():
      indices = {value: i for i, value in enumerate(domain)}
      encoded = [indices[row[key]] for row in padded]
      self.outputs[key] = tuple(build(tuple(bool(value & (1 << bit)) for value in encoded), self.width - 1)
                                for bit in range((len(domain) - 1).bit_length()))

  def read(self, bits):
    bits = tuple(bits)
    if len(bits) != self.width:
      raise ValueError("ROM address width does not match its finite table")
    expressions = [FALSE, TRUE]
    for bit, yes, no in self.nodes:
      expressions.append(choose(bits[bit], expressions[yes], expressions[no]))
    return {key: Value.encoded(self.domains[key], (expressions[index] for index in outputs))
            for key, outputs in self.outputs.items()}


def controller_table(program, names, readers=None, reader_names=(), *, batched=False, augment=None):
  """Decode a GS finite table, optionally including colored data-head roles."""
  tests = {"equal", "less", "symbols", "available"}
  rows = []
  for state, (event, targets) in enumerate(program.code):
    op, *args = event
    row = dict(op=op, left=names[0], right=names[0], direction=1,
               bit=False, no=targets[0] if targets else state,
               yes=targets[1] if op in tests else targets[0] if targets else state)
    if op in ("equal", "less", "symbols", "assert_equal", "copy"):
      row["left"], row["right"] = args
    elif op == "move":
      if not batched and (len(args[0]) != 1 or abs(args[0][0][1]) != 1):
        raise ValueError("unit movement table required")
      if not batched:
        row["left"], row["direction"] = args[0][0]
    elif op in ("available", "match", "border"):
      row["left"] = args[0]
    elif op == "flag":
      row["bit"] = args[0]
    elif op != "halt":
      raise ValueError("unknown finite GS instruction")
    if readers is not None:
      row.update(data_left=None, data_right=None, data_target=None, data_source=None, data_move=None)
      if batched:
        row.update({"data_delta." + name: 0 for name in reader_names})
      color = lambda name: reader_names[readers.colors[name]]
      if op == "symbols":
        row["data_left"], row["data_right"] = map(color, args)
      elif op == "available":
        row["data_left"] = color(args[0])
      elif op == "copy" and args[0] in readers.after[state]:
        row["data_target"], row["data_source"] = map(color, args)
      elif op == "move":
        if batched:
          for head, delta in args[0]:
            if head in readers.after[state]:
              row["data_delta." + color(head)] += delta
        elif row["left"] in readers.after[state]:
          row["data_move"] = color(row["left"])
    rows.append(row)
  domains = dict(op=tuple(dict.fromkeys(row["op"] for row in rows)), left=tuple(names), right=tuple(names),
                 direction=(-1, 1), bit=(False, True), no=tuple(range(len(rows))), yes=tuple(range(len(rows))))
  if readers is not None:
    domains.update({key: (None, *reader_names) for key in
                    ("data_left", "data_right", "data_target", "data_source", "data_move")})
    if batched:
      for name in reader_names:
        key = "data_delta." + name
        domains[key] = tuple(dict.fromkeys(row[key] for row in rows))
  if augment is not None:
    augment(rows, domains)
  return ROM(rows, domains)

File: scala/pal/src/main/scala/pal/ScaffoldRom.scala
package pal

import scala.collection.immutable.VectorMap
import scala.collection.mutable

/** Finite lookup tables as shared Boolean decision diagrams.
  * A read instantiates a fixed Boolean DAG; emitted PEGs contain no array lookup.
  * Port of `scaffold_rom.py`.
  */
final class ROM(sourceRows: Iterable[collection.Map[String, Any]],
                sourceDomains: collection.Map[String, Seq[Any]]) {
  val rows: Vector[VectorMap[String, Any]] = sourceRows.iterator.map(VectorMap.from).toVector
  val domains: VectorMap[String, Vector[Any]] = VectorMap.from(sourceDomains.iterator.map {
    case (key, values) => key -> values.toVector
  })
  if (rows.isEmpty || rows.exists(_.keySet != domains.keySet)) {
    throw new IllegalArgumentException("every finite table row must cover the declared columns")
  }
  if (domains.values.exists(d => d.isEmpty || d.distinct.size != d.size)) {
    throw new IllegalArgumentException("finite column values must be distinct")
  }
  val width: Int = 32 - Integer.numberOfLeadingZeros(rows.size - 1)
  private val nodeBuffer = mutable.ArrayBuffer.empty[(Int, Int, Int)]
  private val unique = mutable.HashMap.empty[(Int, Int, Int), Int]
  private val memo = mutable.HashMap.empty[(Int, Vector[Boolean]), Int]

  private def build(values: Vector[Boolean], bit: Int): Int = {
    if (!values.contains(true)) { 0 }
    else if (!values.contains(false)) { 1 }
    else {
      memo.getOrElseUpdate((bit, values), {
        val half = values.size / 2
        val no = build(values.take(half), bit - 1)
        val yes = build(values.drop(half), bit - 1)
        if (yes == no) { yes }
        else {
          val node = (bit, yes, no)
          unique.getOrElseUpdate(node, {
            val index = nodeBuffer.size + 2
            nodeBuffer += node
            index
          })
        }
      })
    }
  }

  val outputs: VectorMap[String, Vector[Int]] = {
    val padded = rows ++ Vector.fill((1 << width) - rows.size)(rows.head)
    VectorMap.from(domains.iterator.map { case (key, domain) =>
      val indices = domain.zipWithIndex.toMap
      val encoded = padded.map(row => indices(row(key)))
      val bits = 32 - Integer.numberOfLeadingZeros(domain.size - 1)
      key -> Vector.tabulate(bits)(bit => build(encoded.map(v => (v & (1 << bit)) != 0), width - 1))
    })
  }
  val nodes: Vector[(Int, Int, Int)] = nodeBuffer.toVector

  def read(bits: Seq[Expr]): VectorMap[String, Value[Any]] = {
    if (bits.size != width) {
      throw new IllegalArgumentException("ROM address width does not match its finite table")
    }
    val expressions = mutable.ArrayBuffer[Expr](Expr.FALSE, Expr.TRUE)
    nodes.foreach { case (bit, yes, no) =>
      expressions += ScaffoldCircuit.choose(bits(bit), expressions(yes), expressions(no))
    }
    VectorMap.from(outputs.iterator.map { case (key, indices) =>
      key -> Value.encoded(domains(key), indices.map(expressions))
    })
  }
}

object ScaffoldRom {
  type TableRow = mutable.LinkedHashMap[String, Any]
  type Domains = mutable.LinkedHashMap[String, Seq[Any]]

  /** Decode a GS finite table, optionally including colored data-head roles. */
  def controllerTable(program: Program, names: Seq[String], readers: Option[ReaderLiveness] = None,
                      readerNames: Seq[String] = Vector.empty, batched: Boolean = false,
                      augment: Option[(Vector[TableRow], Domains) => Unit] = None): ROM = {
    import Event.*
    val tests = Set("equal", "less", "symbols", "available")
    val rows = program.code.zipWithIndex.map { case (instruction, state) =>
      val event = instruction.event
      val next = instruction.targets.headOption.getOrElse(state)
      val row = mutable.LinkedHashMap[String, Any](
        "op" -> event.op, "left" -> names.head, "right" -> names.head, "direction" -> 1,
        "bit" -> false, "no" -> next,
        "yes" -> (if (tests(event.op)) instruction.targets(1) else next))
      def pair(left: String, right: String): Unit = { row("left") = left; row("right") = right }
      event match {
        case Equal(l, r) => pair(l, r)
        case Less(l, r) => pair(l, r)
        case Symbols(l, r) => pair(l, r)
        case AssertEqual(l, r) => pair(l, r)
        case Copy(l, r) => pair(l, r)
        case Move(moves) =>
          if (!batched && (moves.size != 1 || math.abs(moves.head.delta) != 1)) {
            throw new IllegalArgumentException("unit movement table required")
          }
          if (!batched) { row("left") = moves.head.head; row("direction") = moves.head.delta }
        case Available(head) => row("left") = head
        case Match(head) => row("left") = head
        case Border(head) => row("left") = head
        case Flag(bit) => row("bit") = bit
        case Halt => ()
      }
      readers.foreach { live =>
        Vector("data_left", "data_right", "data_target", "data_source", "data_move").foreach(row(_) = None)
        if (batched) { readerNames.foreach(name => row("data_delta." + name) = 0) }
        def color(name: String): String = readerNames(live.colors(name))
        event match {
          case Symbols(l, r) => row("data_left") = color(l); row("data_right") = color(r)
          case Available(head) => row("data_left") = color(head)
          case Copy(target, source) if live.after(state)(target) =>
            row("data_target") = color(target); row("data_source") = color(source)
          case Move(moves) if batched =>
            moves.filter(m => live.after(state)(m.head)).foreach { move =>
              val key = "data_delta." + color(move.head)
              row(key) = row(key).asInstanceOf[Int] + move.delta
            }
          case Move(moves) if live.after(state)(moves.head.head) => row("data_move") = color(moves.head.head)
          case _ => ()
        }
      }
      row
    }
    val domains: Domains = mutable.LinkedHashMap(
      "op" -> rows.map(_("op")).distinct, "left" -> names, "right" -> names,
      "direction" -> Vector(-1, 1), "bit" -> Vector(false, true),
      "no" -> rows.indices.toVector, "yes" -> rows.indices.toVector)
    if (readers.nonEmpty) {
      Vector("data_left", "data_right", "data_target", "data_source", "data_move").foreach { key =>
        domains(key) = Vector(None) ++ readerNames
      }
      if (batched) {
        readerNames.foreach { name =>
          val key = "data_delta." + name
          domains(key) = rows.map(_(key)).distinct
        }
      }
    }
    augment.foreach(_(rows, domains))
    new ROM(rows, domains)
  }
}
